### ISSUE TITLE
[modbus] 2.4 binding crashes during swap float32

### ISSUE THREAD

**Issue by IrisCrimson** [openhab/openhab-addons#5476]
modbus stops reading registerdata

modbus 2.4

Here the error trace:
```
2019-04-17 12:25:32.779 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.NumberFormatException: null
        at java.math.BigDecimal.<init>(BigDecimal.java:494) ~[?:?]
        at java.math.BigDecimal.<init>(BigDecimal.java:383) ~[?:?]
        at java.math.BigDecimal.<init>(BigDecimal.java:806) ~[?:?]
        at java.math.BigDecimal.valueOf(BigDecimal.java:1274) ~[?:?]
        at org.eclipse.smarthome.core.library.types.DecimalType.<init>(DecimalType.java:53) ~[?:?]
        at org.openhab.io.transport.modbus.ModbusBitUtilities.extractStateFromRegisters(ModbusBitUtilities.java:172) ~[?:?]
        at org.openhab.binding.modbus.internal.handler.ModbusDataThingHandler.onRegisters(ModbusDataThingHandler.java:664) ~[?:?]
        at org.openhab.binding.modbus.internal.handler.ModbusPollerThingHandlerImpl$ReadCallbackDelegator.lambda$0(ModbusPollerThingHandlerImpl.java:91) ~[?:?]
        at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[?:?]
        at org.openhab.binding.modbus.internal.handler.ModbusPollerThingHandlerImpl$ReadCallbackDelegator.onRegisters(ModbusPollerThingHandlerImpl.java:91) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusLibraryWrapper.invokeCallbackWithResponse(ModbusLibraryWrapper.java:279) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl$PollOperation.lambda$1(ModbusManagerImpl.java:163) ~[?:?]
        at org.openhab.io.transport.modbus.internal.SimpleStopWatch.timeRunnable(SimpleStopWatch.java:148) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl$PollOperation.accept(ModbusManagerImpl.java:162) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl$PollOperation.accept(ModbusManagerImpl.java:1) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl.executeOperation(ModbusManagerImpl.java:571) ~[?:?]
        at org.openhab.io.transport.modbus.internal.ModbusManagerImpl.lambda$15(ModbusManagerImpl.java:719) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
```

## Expected Behavior

If the readed register does not contain valid data, or it could not be read by any reason,
the binding should be robust against such corruped data. Right now the thread is crashing and reading of the registers is stopped.

catch the error and try it again next cycle.  After e.g. 10 invalid reads in a row set an error.

## Possible Solution
catch the error and try it again next cycle.  After e.g. 10 invalid reads in a row set an error.

## Steps to Reproduce (for Bugs)
sporadic -> maybe my wago SPS does not always like the floats

## Context
recording of temp data is interupted an I have to monitor it manualy if binding stopped working.
With the modbus 1.x binding it was running without problems 1 and a half year.

## Your Environment
openhabian 2.4

**Comment by ssalonen** [comment]
Can you please provide thing definitions and item definitions?

**Comment by ssalonen** [comment]
Based on stack trace I cannot think of anything else except registers having floating point number not-a-number or infinity. DecimalType cannot handle/represent these. 

I think correct fix would be to use UNDEF in these situations.

**Comment by ssalonen** [comment]
Pull request submitted. Nan or infinity floating point values are translated to UNDEF.

**Comment by IrisCrimson** [comment]
Hi Sami

Sorry for my late response, had many family easter events....

Here the thing definition:

//starting from address 60 sensor data 
	Bridge poller wago_slave3 [ start=12348, length=60, refresh=500, type="holding" ] {
		/* old binding config
tcp.slave3.connection=192.168.178.55:502
tcp.slave3.type=holding
tcp.slave3.start=12348
tcp.slave3.length=60
tcp.slave3.valuetype=float32_swap		 * 
		 */
		Thing data wago_s3_000 [ readStart="12348", readValueType="float32_swap" ]
		Thing data wago_s3_001 [ readStart="12350", readValueType="float32_swap" ]
		Thing data wago_s3_002 [ readStart="12352", readValueType="float32_swap" ]
		Thing data wago_s3_003 [ readStart="12354", readValueType="float32_swap" ]
		Thing data wago_s3_004 [ readStart="12356", readValueType="float32_swap" ]
		Thing data wago_s3_005 [ readStart="12358", readValueType="float32_swap" ]
		Thing data wago_s3_006 [ readStart="12360", readValueType="float32_swap" ]
		Thing data wago_s3_007 [ readStart="12362", readValueType="float32_swap" ]
		Thing data wago_s3_008 [ readStart="12364", readValueType="float32_swap" ]
		Thing data wago_s3_009 [ readStart="12366", readValueType="float32_swap" ]
		Thing data wago_s3_010 [ readStart="12368", readValueType="float32_swap" ]
		Thing data wago_s3_011 [ readStart="12370", readValueType="float32_swap" ]
		Thing data wago_s3_012 [ readStart="12372", readValueType="float32_swap" ]
		Thing data wago_s3_013 [ readStart="12374", readValueType="float32_swap" ]
		Thing data wago_s3_014 [ readStart="12376", readValueType="float32_swap" ]
		Thing data wago_s3_015 [ readStart="12378", readValueType="float32_swap" ]
		Thing data wago_s3_016 [ readStart="12380", readValueType="float32_swap" ]
		Thing data wago_s3_017 [ readStart="12382", readValueType="float32_swap" ]
		Thing data wago_s3_018 [ readStart="12384", readValueType="float32_swap" ]
		Thing data wago_s3_019 [ readStart="12386", readValueType="float32_swap" ]
		Thing data wago_s3_020 [ readStart="12388", readValueType="float32_swap" ]
		Thing data wago_s3_021 [ readStart="12390", readValueType="float32_swap" ]
		Thing data wago_s3_022 [ readStart="12392", readValueType="float32_swap" ]
		Thing data wago_s3_023 [ readStart="12394", readValueType="float32_swap" ]
		Thing data wago_s3_024 [ readStart="12396", readValueType="float32_swap" ]
/*		Thing data wago_s3_025 [ readStart="12398", readValueType="float32_swap" ]
		Thing data wago_s3_026 [ readStart="12400", readValueType="float32_swap" ]
		*/
		Thing data wago_s3_027 [ readStart="12402", readValueType="float32_swap" ]

/*
	
		Thing data wago_s3_028 [ readStart="12404", readValueType="float32_swap" ]
		Thing data wago_s3_029 [ readStart="12406", readValueType="float32_swap" ]
*/
	}


And here some of the items:

Number PufferT1 "Puffer Temp T1" (gKessel) {channel="modbus:data:wago:wago_slave3:wago_s3_000:number"}
Number PufferT2 "Puffer Temp T2" (gKessel) {channel="modbus:data:wago:wago_slave3:wago_s3_001:number"}
Number PufferT3 "Puffer Temp T3" (gKessel) {channel="modbus:data:wago:wago_slave3:wago_s3_002:number"}
Number PufferT4 "Puffer Temp T4" (gKessel) {channel="modbus:data:wago:wago_slave3:wago_s3_003:number"}

To set it direclty to UNDEF in this case (NaN, Inf, ..) I'm not 100% sure if this is the best solution, but as long as the thread is not chrashing anymore your fix is really great.

My opinion is still the thread should try to read it at least once (or 10 times) more, and if the value is still invalid then set it to UNDEF. Otherwise keep the old value. Main reason is that the Modbus connection/protocol and the data is not protected somehow.

But for now I'm really happy that you have fixed it so fast.

Thanks :)

**Comment by ssalonen** [comment]
Thanks! If you can get verbose logs when the error happens, that would be great.

Regarding your proposal to have special retries with NaN/inf values, I respectfully disagree. 

What we now have is just an oversight the way floating point special values are handled in the binding. Ideally we would like to pass NaN/infinity to channels even, but since that is not possible (DecimalType is designed to work with regular numbers only), UNDEF/null will have to do.

Since modbus binding use cases are so varied, I don't think we can assume that NaN means "device temporarily broken". I would argue that in some cases it is relevant to get the information when we receive NaN value.

In other words, I think there is nothing wrong with the data received, it just happens to refer special value in IEE 754 floating point spec.

I doubt that it is "data corruption" in the protocol, there's many layers of protection against that (in your case tcp, in addition to modbus protocol check sums).

However, as you also pointed out data will be regularly polled, in case you have poller defined. So if the NaN value is some sort of transient error with your device, it will be retried automatically. You can even request refresh on UNDEF values (check out eager refresh example in the docs)
